lib/memory.c

```diff
@@ -112,6 +119,102 @@ static void alloc_tail(void *area, size_t size)
 	}
 }
 
+#if USE_ALLOC_PROF
+
+M0_INTERNAL void *m0_alloc_nz(size_t size)
+{
+	return m0_alloc_profiled(size, M0_ALLOC_CALLSITE("alloc-nz", 0));
+}
+
+void *m0_alloc(size_t size)
+{
+	return m0_alloc_profiled(size, M0_ALLOC_CALLSITE("alloc", 0));
+}
+
+M0_INTERNAL void *m0_alloc_aligned(size_t size, unsigned shift)
+{
+	return m0_alloc_aligned_profiled(size, shift,
+					 M0_ALLOC_CALLSITE("alloc-aligned", 0));
+}
+
+struct pad {
+	struct m0_alloc_callsite *p_cs;
+	uint64_t                  p_magic;
+};
+
+void *m0_alloc_profiled(size_t size, struct m0_alloc_callsite *cs)
+{
+	struct pad *area = alloc0(size + sizeof(struct pad));
```

Comment:
I suggest to have:
```
struct pad *area = alloc0(sizeof(struct pad) + size);
```
This gives the layout that the 'pad' is at the beginning and the actual content follows.

Reply:
pad structure is a place holder for allocation and free stats, @nikitadanilov  if we change/swap the offset, will it create any issue ?

Reply:
Wait... the pad is already at the beginning of the allocated area. `sizeof(struct pad) + size` is just size calculation it does not specify any layout. @huanghua78, I am confused.

Reply:
Maybe I confused you too.
`sizeof(struct pad) + size` is the size calculation.
And I know the pad is at the beginning of the allocated area.

If you compare the `size + sizeof(struct pad)` and the `sizeof(struct pad) + size`, the results are the same.
This is the size of allocated area, including pad at the beginning.
I know that.
`sizeof(struct pad) + size` will give us an impression that the 'pad' goes first, and payload goes next.


If I've created any confusion, I am sorry, and just skip this comment.